src/demikernel/bindings.rs

```diff
@@ -332,20 +451,76 @@ pub extern "C" fn dmtr_sgaalloc(size: libc::size_t) -> dmtr_sgarray_t {
 #[no_mangle]
 pub extern "C" fn dmtr_sgafree(sga: *mut dmtr_sgarray_t) -> c_int {
     trace!("dmtr_sgfree()");
+
+    // Check if scatter-gather array is invalid.
     if sga.is_null() {
-        return 0;
+        return libc::EINVAL;
     }
-    with_libos(|libos| {
-        libos.sgafree(unsafe { *sga }).unwrap();
-        0
+
+    // Issue sgafree operation.
+    with_libos(|libos| match libos.sgafree(unsafe { *sga }) {
+        Ok(()) => 0,
+        Err(e) => {
+            warn!("sgafree() failed: {:?}", e);
+            e.errno
+        },
     })
 }
 
 //==============================================================================
 // getsockname
 //==============================================================================
 
+#[allow(unused)]
 #[no_mangle]
 pub extern "C" fn dmtr_getsockname(qd: c_int, saddr: *mut sockaddr, size: *mut socklen_t) -> c_int {
-    unimplemented!()
+    // TODO: Implement this system call.
+    libc::ENOSYS
+}
+
+//==============================================================================
+// setsockopt
+//==============================================================================
+
+#[allow(unused)]
+#[no_mangle]
+pub extern "C" fn dmtr_setsockopt(
+    qd: c_int,
+    level: c_int,
+    optname: c_int,
+    optval: *const c_void,
+    optlen: socklen_t,
+) -> c_int {
+    // TODO: Implement this system call.
+    libc::ENOSYS
+}
+
+//==============================================================================
+// getsockopt
+//==============================================================================
+
+#[allow(unused)]
+#[no_mangle]
+pub extern "C" fn dmtr_getsockopt(
+    qd: c_int,
+    level: c_int,
+    optname: c_int,
+    optval: *mut c_void,
+    optlen: *mut socklen_t,
+) -> c_int {
+    // TODO: Implement this system call.
+    libc::ENOSYS
+}
+
+//==============================================================================
+// Standalone Functions
+//==============================================================================
+
+/// Converts a [sockaddr] into a [Ipv4Endpoint].
+fn sockaddr_to_ipv4endpoint(saddr: *const sockaddr) -> Result<Ipv4Endpoint, Fail> {
+    let sin: libc::sockaddr_in =
+        unsafe { *mem::transmute::<*const sockaddr, *const libc::sockaddr_in>(saddr) };
+    let addr: Ipv4Addr = { Ipv4Addr::from(u32::from_be_bytes(sin.sin_addr.s_addr.to_le_bytes())) };
```

Comment:
What is going on here with byte order?  To the naive reader (i.e. me :-)) this looks like s_addr is being taken as little endian and being read as big endian.

Reply:
I did not author this and I also have no clue why the person intentionally did it.

I just noticed that the same code was everywhere, so I at least created a function to do this copy-and-paste job.

I'm happy to take any improvements on this function afterwards, but considering its name, please let me know what your suggestion is.

Reply:
Okay, mark it with a ToDo comment to review it later, file an Issue about it, and we'll leave it as is for now.  My guess is that whoever wrote this didn't understand what was going on and just kept putting in byte-order conversions until it worked :-).

Reply:
Sure. I created issue https://github.com/demikernel/demikernel/issues/101 to keep track of this.